### PR TITLE
Adding test for leak detection and fixing up build to not leak

### LINKED ISSUE #46: Propel is leaking in Global Scope
TL;DR Propel is leaking a tonne of variables on the global scope and Mocha wasn't catching it due to the formation of the tests

Mocha has leak detection, but the way it works is it compares the current state *before* the test to the state *after* the test.

The problem with this is that the mocha test page loads the script in the html (i.e. before the tests run)

    <script src="/dist/client.js"></script>

This means the leaks aren't treated as leaks. The fix for the test is to add a test that asynchronously loads the script and this will trigger the leak detection.

      it('should load window.goog.propel.Client without leaks', done => {
        // By leaks this is referring to the only thing Propel
        // should add to the global scope (i.e. window) is goog
        const scriptElement = document.createElement('script');
        scriptElement.setAttribute('type', 'text/javascript');
        scriptElement.src = '/dist/client.js';
        document.querySelector('head').appendChild(scriptElement)
        .onload = () => {
          console.log('Script Loaded');
          done();
        };
      });

### PR DESCRIPTION
**Pull request by gauntface**
This fixes #46 and #32 

@petele for naming
@surma for build process
@wibblymat as FYI

### CODE REVIEW

File: gulp-tasks/build.js [hunk 1]
```diff
@@ -18,37 +18,64 @@
 'use strict';
 
 const gulp = require('gulp');
-const babel = require('gulp-babel');
 const uglify = require('gulp-uglify');
-const rollup = require('gulp-rollup');
 const sourcemaps = require('gulp-sourcemaps');
-const commonjs = require('rollup-plugin-commonjs');
-const nodeResolve = require('rollup-plugin-node-resolve');
+const browserify = require('browserify');
```
Comment (surma) [inline gulp-tasks/build.js]: ... hello darkness my old friend

File: gulp-tasks/build.js [hunk 2]
```diff
@@ -18,37 +18,64 @@
...
+  });
+
+  // Rollupify reduces the size of the final output but increases build
+  // time to do it so enable for production build only
+  if (GLOBAL.config.env === 'prod') {
+    browserifyBundles.transform('rollupify');
+  }
+
+  let stream = browserifyBundles
+  .transform('babelify', {presets: ['es2015']})
```
Comment (surma) [inline gulp-tasks/build.js]: Your indentation is screwed up here

File: test/browser-tests/push-client/push-client-constructor.js [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+// This is a test and we want descriptions to be useful, if this
+// breaks the max-length, it's ok.
+
+/* eslint-disable max-len, no-unused-expressions */
+/* eslint-env browser, mocha */
+
+'use strict';
+
+describe('Test PushClient Constructor', () => {
+  it('should load window.goog.propel.Client without leaks', done => {
```
Comment (surma) [inline test/browser-tests/push-client/push-client-constructor.js]: Where are you checking for leaks here?

Comment (gauntface) [inline test/browser-tests/push-client/push-client-constructor.js]: It occurs in index.html with the call to mocha.checkLeaks()

File: gulp-tasks/build.js (hunk 2 above)
Comment (gauntface) [inline gulp-tasks/build.js]: what do you mean? I'm missing something sorry :(

Comment (surma) [inline gulp-tasks/build.js]: `.transform` and every subsequent dot should be indented, shouldn’t they?

File: test/browser-tests/push-client/push-client-constructor.js (hunk 3 above)
Comment (surma) [inline test/browser-tests/push-client/push-client-constructor.js]: I am smart.

**Pull request merged**